### DISCUSSION TITLE
uasyncio: cancel a task which is running inside a task

### DISCUSSION THREAD

**Discussion by rootfrogsMakes** [micropython/micropython#13431] [Using MicroPython] [answered]
Hello,
I deep dived a little bit into multi tasking with uasyncio. I am observing behaviour that I cannot explain.
Maybe some one know whats going wrong?

To show the issue more easy I wrote a small example:

```python
import uasyncio

async def myPrinter():
    global subTask
    while(True):
        print("== Printing now ==")
        subTask = uasyncio.create_task(subPrinter())
        try:
            await uasyncio.wait_for(subTask, 10000)
        finally:
            print("Subtask finished")
        print("== Printing finished ==")
		
async def subPrinter():
    for i in range(7):
        print("sub print")
        await uasyncio.sleep_ms(500)

async def mainLoop():
    global subTask
    printerTask = uasyncio.create_task(myPrinter())
    await uasyncio.sleep_ms(2000)
    print("Its time to stop subPrinter...")
    subTask.cancel()
    print("SubPrinter is stopped")
    await uasyncio.sleep_ms(10000)
    print("Program is finished")

uasyncio.run(mainLoop())
```

So I start a ``printerTask`` with function ``myPrinter``. Inside the ``printerTask`` I create a ``subTask``. Its Task-Object is global.
In the Main loop I now try to stop the ``subTask`` by calling ``subTask.cancel()``. The task cancels (instead of 7 times printing "sub print", It just prints it 4 times) and I get the printout "Subtask finished".
However, not only the ``subTask`` is cancelled, but also the ``printerTask``. The finally statement which prints out "Subtask finished" gets called, but the command after it with the printout "== Printing finished ==" isn't called. So it seems that the printerTask also get cancelled.

Why can't I just stop the ``subTask``?  Is it possible to cancel just the ``subTask`` and keeps the ``printerTask`` running?

Output of the program above:
```
== Printing now ==
sub print
sub print
sub print
sub print
Its time to stop subPrinter...
SubPrinter is stopped
Subtask finished
Program is finished
```

**Comment by ned-pcs** [comment]
Add another `try/except`:
```python
async def myPrinter():
    global subTask
    while True:
        print("== Printing now ==")
        try:
            subTask = uasyncio.create_task(subPrinter())
            try:
                await uasyncio.wait_for(subTask, 10000)
            finally:
                print("Subtask finished")
            print("== Printing finished ==")
        except uasyncio.CancelledError:
            print("Subtask cancelled")
```

On my system (using the Linux build of Micropython) I get this output:
```
== Printing now ==
sub print
sub print
sub print
sub print
Its time to stop subPrinter...
SubPrinter is stopped
Subtask finished
Subtask cancelled
== Printing now ==
sub print
sub print
sub print
sub print
sub print
sub print
sub print
Subtask finished
== Printing finished ==
== Printing now ==
sub print
sub print
sub print
sub print
sub print
sub print
sub print
Subtask finished
== Printing finished ==
== Printing now ==
sub print
sub print
sub print
sub print
sub print
sub print
Program is finished
```

**Comment by rootfrogsMakes** [comment, marked as answer]
Oh, thanks for the hint.
You are right, I forgot the except statement. It also works if I change the finally statement to an except path:

```python
async def myPrinter():
    global subTask
    while(True):
        print("== Printing now ==")
        subTask = uasyncio.create_task(subPrinter())
        try:
            await uasyncio.wait_for(subTask, 10000)
        except:
            print("subtask exception")
```

Due to the missing exception handling for the subTask, the ``uasyncio.CancelledError`` will affect the printerTask too. But if I handle it, the printerTask keeps running.

Thank you for the idea.
